core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -186,6 +186,11 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         // automatically, so try to tell the executor to stop itself. See SPARK-13519.
         executorDataMap.get(executorId).foreach(_.executorEndpoint.send(StopExecutor))
         removeExecutor(executorId, reason)
+
+      case ExecutorConstructed(executorId) =>
+        executorDataMap.get(executorId).foreach { data =>
+          data.freeCores = data.totalCores
+        }
```

Comment:
Don't you need to call `makeOffers(executorId)` to have the executor get tasks right away, instead of waiting for other events to trigger that?

Simiarly, the old "register" message handler does not need to call `makeOffers` anymore.

Reply:
good catch, we shall move the call of `makeOffers` from `RegisterExecutor` to here.